src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+                c.RegisterOperationAction(ctx => coll.HandleVariableDeclarator((IVariableDeclaratorOperation)ctx.Operation), OperationKind.VariableDeclarator);
+                c.RegisterOperationAction(ctx => coll.HandleDeclarationExpression((IDeclarationExpressionOperation)ctx.Operation), OperationKind.DeclarationExpression);
+                c.RegisterOperationAction(ctx => coll.HandleReturn((IReturnOperation)ctx.Operation), OperationKind.Return);
+                c.RegisterCompilationEndAction(ctx => Report(ctx, state));
+            });
+        }
+
+        /// <summary>
+        /// Given all the accumulated analysis state, generate the diagnostics.
+        /// </summary>
+        private static void Report(CompilationAnalysisContext context, State state)
+        {
+            var types = new HashSet<ITypeSymbol>(SymbolEqualityComparer.Default);
+
+            foreach (var field in state.VirtualDispatchFields.Keys)
+            {
+                if (field.DeclaredAccessibility == Accessibility.Public
+                    || field.DeclaredAccessibility == Accessibility.Internal
+                    || field.DeclaredAccessibility == Accessibility.ProtectedOrInternal)
+                {
+                    continue;
+                }
+
+                if (state.FieldAssignments.TryGetValue(field, out var assignments))
+                {
+                    Report(field, field.Type, assignments, UseConcreteTypeForField);
+                }
+            }
+
+            foreach (var local in state.VirtualDispatchLocals.Keys)
+            {
+                if (state.LocalAssignments.TryGetValue(local, out var assignments))
+                {
+                    Report(local, local.Type, assignments, UseConcreteTypeForLocal);
+                }
+            }
+
+            foreach (var parameter in state.VirtualDispatchParameters.Keys)
+            {
+                if (state.ParameterAssignments.TryGetValue(parameter, out var assignments))
+                {
+                    if (parameter.ContainingSymbol is IMethodSymbol method)
+                    {
+                        if (CanUpgrade(method))
+                        {
+                            Report(parameter, parameter.Type, assignments, UseConcreteTypeForParameter);
+                        }
+                    }
+                }
+            }
+
+            foreach (var pair in state.MethodReturns)
+            {
+                var method = pair.Key;
+                var returns = pair.Value;
+
+                if (CanUpgrade(method))
+                {
+                    Report(method, method.ReturnType, returns, UseConcreteTypeForMethodReturn);
+                }
+            }
+
+            void Report(ISymbol sym, ITypeSymbol fromType, ConcurrentBag<ITypeSymbol> assignments, DiagnosticDescriptor desc)
+            {
+                types.Clear();
```

Comment:
Instead of using the same hashset across method calls to the Report method, can we just use a `PooledHashSet` that is available in the repo? Please see other use sites in the repo for the usage pattern of `PooledHashSet`.

Reply:
Done.